Document how to create/update an index snapshot. (#243)
optakt/flow-dps: docs/snapshots.md

> # Index snapshots

Reviewer: ```suggestion
# Index Snapshots
```

See https://en.wikipedia.org/wiki/Title_case#Chicago_Manual_of_Style

Author: Sorry, I must be so annoying 😅 This is at least the third time you're correcting me for the same thing. Fixing!

This document describes index snapshots, what they are and how they can be created or updated.

> **Table of Contents**
> - [What are index snapshots](#what-are-snapshots)
> - [Create a snapshot](#create-a-snapshot)
> - [Updating a snapshot](#update-a-snapshot)
>     1. [Create a new index](#create-a-new-index)
>     2. [Create an index snapshot](#create-an-index-snapshot)

Reviewer: ```suggestion
**Table of Contents**

- [What are index snapshots](#what-are-snapshots)
- [Creating a snapshot](#creating-a-snapshot)
- [Updating a snapshot](#updating-a-snapshot)
    1. [New index](#new-index)
    2. [New snapshot](#new-snapshot)
```

WDYT?

Author: Looks good, opted for that variant :)

> ## What are index snapshots

Reviewer: ```suggestion
## What Are Index Snapshots
```

Index snapshots are images of the DPS index at a certain point in time.
These images can be used to create or restore the content of an index in order to operate on a certain indexed data.
Index snapshots can be used to create a starting point for tests, so that tests have actual blocks, accounts, transactions and other information to operate on.

At a low level, snapshots are created using the [badger](https://github.com/dgraph-io/badger) backup functionality.
Technical documentation can be found [here](https://pkg.go.dev/github.com/dgraph-io/badger/v3#DB.Backup).

> ## Create a snapshot

Reviewer: ```suggestion
## Creating a Snapshot
```

Index snapshots are created using the CLI tool found at `cmd/create-index-snapshot`.
Usage for the tool is described in more detail [here](https://github.com/optakt/flow-dps/blob/master/cmd/create-index-snapshot/README.md).

> ## Update a snapshot

Reviewer: ```suggestion
## Updating a snapshot
```

> It is sometimes necessary to regenerate the snapshot.
> Perhaps in order to add additional blocks, events or some other data to the snapshot to be referenced in tests.

Reviewer: ```suggestion
It is sometimes necessary to regenerate the snapshot, for example when an index was added or changed.
```

WDYT?

Other reasons may be because the internal format of the index changed, like when a compression dictionary is updated.
When an index snapshot is created, it is compressed using a specific compression dictionary.

> When restoring the index, the snapshot needs to be decompressed using the same dictionary or decompression will fail.

Reviewer: Seems strange that the first sentence of this paragraph is related to the previous paragraph. The others are not 🤔 I think it should be with the previous paragraph, and the two next sentences should remain here.

> ### Create a new index

Reviewer: ```suggestion
### New Index
```

> Create a new index using the same inputs as for the original index.

Reviewer: ```suggestion
The following command creates a new index. Be careful to always use the same inputs as for the original index.
```

Currently what's the way to get this data? The only way is to ask you for the tarball? :p

Author: The doc was intended to be generic, since I assume we will have different snapshots for different scenarios.

I think it was not the plan to keep iterating on the same thing repeatedly, and snapshot was supposed to be the source of truth once created. Else we would just have the protocol state archive versioned too. Not sure though

```console
$ flow-dps-indexer -t <trie_dir> -d <data_dir> -i <index_dir> -a
```

> ### Create an index snapshot

Reviewer: ```suggestion
### New Snapshot
```

Creating an index snapshot is the same as described in the [Create a snapshot](#create-a-snapshot) section, while referencing the new index.

```console
$ create-index-snapshot -i <new_index_dir> > output.hex
```

> By default the snapshot is hex-encoded.

Reviewer: It's strange that we have a section `Creating a Snapshot` which redirects to the cmd documentation, and then here we address the exact same question ("how to create a snapshot?") in a different way. In the same document 😅

Author: Yeah, I restructured it a bit, think it's better now. It was a bit weird.
